agreement_rebate/models/agreement_rebate_settlement.py

```diff
@@ -0,0 +1,243 @@
+# Copyright 2020 Tecnativa - Carlos Dauden
+# Copyright 2020 Tecnativa - Sergio Teruel
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import _, api, models, fields
+from odoo.osv import expression
+from odoo.tools import safe_eval
+from odoo.exceptions import UserError
+
+
+class AgreementRebateSettlement(models.Model):
+    _name = 'agreement.rebate.settlement'
+    _description = 'Agreement Rebate Settlement'
+    _order = 'date DESC'
+
+    name = fields.Char()
+    company_id = fields.Many2one(
+        comodel_name='res.company',
+        string='Company',
+        required=True,
+        index=True,
+        default=lambda self: self.env.user.company_id.id,
+    )
+    date = fields.Date(default=fields.Date.today())
+    date_from = fields.Date()
+    date_to = fields.Date()
+    partner_id = fields.Many2one(
+        comodel_name='res.partner',
+        string='Partner',
+    )
+    line_ids = fields.One2many(
+        comodel_name='agreement.rebate.settlement.line',
+        inverse_name='rebate_settlement_id',
+        string='Settlement Lines',
+    )
+    amount_invoiced = fields.Float(string='Amount invoiced')
+    amount_rebate = fields.Float(string='Amount rebate')
+    invoice_id = fields.Many2one(
+        comodel_name='account.invoice',
+        string='Invoice'
+    )
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            if vals.get('name', False):
+                continue
+            vals['name'] = self.env['ir.sequence'].next_by_code(
+                'agreement.rebate.settlement')
+        return super(AgreementRebateSettlement, self).create(vals_list)
+
+    def _prepare_invoice(self):
+        """
+        Prepare the dict of values to create the new invoice for a sales order.
+        This method may be overridden to implement custom invoice generation
+        (making sure to call super() to establish a clean extension chain).
+        """
+        self.ensure_one()
+        company_id = self.company_id.id or self.env.user.company_id.id
+        partner = (self.env.context.get('partner_invoice', False) or
+                   self.partner_id)
+        invoice_type = self.env.context.get('invoice_type', 'out_invoice')
+        journal_id = (
+            self.env.context.get('journal_id') or
+            self.env['account.invoice'].with_context(
+                company_id=company_id
+            ).default_get(['journal_id'])['journal_id'])
+        if not journal_id:
+            raise UserError(_('Please define an accounting sales journal for'
+                              ' this company.'))
+        vinvoice = self.env['account.invoice'].new({
```

Comment:
Please use `odoo.tests.Form()` as standardized since v12 for getting onchange values.